src/api/dirmonitor/polling.c

```diff
@@ -0,0 +1,123 @@
+#include <stdlib.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <SDL.h>
+
+enum {
+  EVENT_MODIFIED
+};
+
+struct file_info {
+  int fd;
+  struct stat file_stat;
+  int events;
+};
+
+struct dirmonitor_internal {
+  struct file_info* fds;
+  size_t fd_count;
+};
+
+  
+struct dirmonitor_internal* init_dirmonitor() {
+  struct dirmonitor_internal* monitor = calloc(1, sizeof(struct dirmonitor_internal));
+
+  monitor->fd_count = 0;
+  monitor->fds = NULL;
+  return monitor;
+}
+
+void deinit_dirmonitor(struct dirmonitor_internal* monitor) {
+  for (size_t i = 0; i < monitor->fd_count; ++i)
+    close(monitor->fds[i].fd);
+  free(monitor->fds);
+}
+
+int get_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int len)
+{
+  int c = 0;
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd < 0)
+      continue;
+
+    struct stat new_stat;
+    fstat(monitor->fds[i].fd, &new_stat);
+
+    if (new_stat.st_mtime > monitor->fds[i].file_stat.st_mtime)
+    {
+      monitor->fds[i].file_stat = new_stat;
+      monitor->fds[i].events = EVENT_MODIFIED;
+      c += 1;
+    }
+  }
+
+  if (!c)
+    SDL_Delay(100);
+
+  return c;
+}
+
+int translate_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int count, int (*change_callback)(int, const char*, void*), void* data) {
+  for (size_t i = 0; i < monitor->fd_count && count; ++i) {
+    if (!(monitor->fds[i].events & EVENT_MODIFIED)) {
+      --count;
+      monitor->fds[i].events = 0;
+      change_callback(monitor->fds[i].fd, NULL, data);
+    }
+  }
+
+  return count != 0;
```

Comment:
```suggestion
  for (size_t i = 0; i < monitor->fd_count; ++i) {
    if (monitor->fds[i].events & EVENT_MODIFIED) {
      monitor->fds[i].events = 0;
      change_callback(monitor->fds[i].fd, NULL, data);
    }
  }

  return 0;
```
I don't think we need to care about `count` as this should just call the callback for every "possible" `fd`.

Also I think the conditional was inverted.

Reply:
The intention behind the code was to just check for as long as we need to, if we know that there are 3 files that were modified we don't need to check 10 more.

The return was to designate that something was incorrect, however the dirmonitor code doesn't assign this return value so every following loop we'd always "error"

Applied your suggestion with some modifications